### ISSUE TITLE
Stream slow starts with HTTP 415 on subtitle track

### ISSUE THREAD

**Issue by dagwieers** [add-ons/plugin.video.vrt.nu#735]
### Describe the bug ###
As reported in #685, the Radio 1 live stream on playback gets a **HTTP 415 Unsupported Media Type**, but after 10 successive attempts it finally starts the stream. **This causes a 10 to 20 seconds delay** in starting the Radio 1 live stream in Kodi.

This vaguely relates to #307 and #727 

### To Reproduce ###
1. Go to **Live TV**
2. Click on **Radio 1 live**
3. Wait 10 seconds
4. See errors in the Kodi log

### Expected behavior ###
Immediate playback like other vualto streams.

### Additional context ###
- Operating system: LibreELEC v9.2.2
- Kodi version: 18.6
- Addon version: 2.3.2
- Using a VPN: no
- Country you are using the addon from: BE

### Log (if available) ###
```paste below
2020-04-11 03:47:40.170 T:1154478976  NOTICE: [plugin.video.vrt.nu] Play: https://live-radio-cf.lwc.vrtcdn.be/groupa/live/d6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4/live.isml/.mpd
2020-04-11 03:47:40.219 T:1937048128  NOTICE: VideoPlayer::OpenFile: plugin://plugin.video.vrt.nu/play/id/vualto_radio1
2020-04-11 03:47:40.310 T:1474552704  NOTICE: Creating InputStream
2020-04-11 03:47:40.410 T:1474552704  NOTICE: Creating Demuxer
2020-04-11 03:47:40.411 T:1474552704  NOTICE: Opening stream: 1001 source: 256
2020-04-11 03:47:40.439 T:1474552704  NOTICE: Creating video codec with codec id: 27
2020-04-11 03:47:40.455 T:1474552704  NOTICE: Opening stream: 1003 source: 256
2020-04-11 03:47:40.483 T:1474552704  NOTICE: Finding audio codec for: 86018
2020-04-11 03:47:40.487 T:1563292544  NOTICE: [plugin.video.vrt.nu] [PlayerInfo 2055] Event onPlayBackStarted
2020-04-11 03:47:40.488 T:1474552704  NOTICE: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
2020-04-11 03:47:40.488 T:1474552704  NOTICE: Opening stream: 1004 source: 256
2020-04-11 03:47:40.489 T:1563292544  NOTICE: [plugin.video.vrt.nu] [PlayerInfo 2055] Avoid setting resumepoints for livestream vualto_radio1
2020-04-11 03:47:40.702 T:1462821760   ERROR: AddOnLog: InputStream Adaptive: Download https://live-radio-cf.lwc.vrtcdn.be/groupa/live/d6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4/live.isml/dash/live-text_track_0_dut=1000-1586569632560.dash failed with error: 415
2020-04-11 03:47:51.219 T:1224029056   ERROR: Previous line repeats 10 times.
2020-04-11 03:47:51.220 T:1224029056  NOTICE: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
2020-04-11 03:47:51.846 T:1310282624  NOTICE: Previous line repeats 1 times.
2020-04-11 03:47:51.846 T:1310282624 WARNING: CRenderManager::WaitForBuffer - timeout waiting for buffer
```

**Comment by dagwieers** [comment]
I reported this to VRT NU, they asked me to report this to the *VRT-Klantendienst* and they answered today:

> We hebben geen gegarandeerde oplossing voor het probleem, vermits we het niet zelf kunnen reproduceren op onze VRT-touchpoints. Daar werken we namelijk met een andere player, die op een andere manier omgaat met de al dan niet meegeleverde data.
>
> Er bleek wel één technisch verschil te zitten tussen het Radio 1-webkanaal en de andere; het heeft te maken met de tijdscode-bron. Het Radio 1-kanaal stond op "media time", in plaats van UNIX timecode. 
>
> We hebben dit gecorrigeerd.

**Comment by dagwieers** [comment]
Adding the detailed insights from @mediaminister from https://github.com/add-ons/plugin.video.vrt.nu/pull/685#issuecomment-580792766 below for convenience:

> Yes,  Unsupported Media Type, there must be something wrong with this stream. I tried playing the stream https://live-radio-cf.lwc.vrtcdn.be/groupa/live/d6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4/live.isml/.mpd with https://www.jwplayer.com/developers/stream-tester/ and I also got an error.
>
> Same 415 error with
> ```
> ffmpeg -i "https://live-radio-cf.lwc.vrtcdn.be/groupa/live/d6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4/live.isml/.mpd"
> ```
>
> ```
> [https @ 0x55a539b87880] HTTP error 415 Unsupported Media Type
> [dash @ 0x55a539a50e00] Failed to open fragment of playlist 0
> ```
>
> VRT should fix this, this error is documented in the [Unified Streaming Platform docs](https://docs.unified-streaming.com/documentation/vod/troubleshooting.html#unsupported-media-type):
>```
>You might experience '415 unsupported media type' errors when a multi-bitrate stream is not keyframe-aligned.
>```
>https://docs.unified-streaming.com/documentation/vod/troubleshooting.html#unsupported-media-type
>
>We experienced this error earlier with some VOD content: https://github.com/pietje666/plugin.video.vrt.nu/issues/307#issuecomment-545425256

**Comment by dagwieers** [comment]
Testing the stream using JW Player, fails playback as well (although it briefly works:
https://flowplayer.com/developers/tools/stream-tester?custom=https%3A%2F%2Flive-radio-cf.lwc.vrtcdn.be%2Fgroupa%2Flive%2Fd6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4%2Flive.isml%2F.mpd

**Comment by mediaminister** [comment]
There is a workaround possible, selecting the HLS-stream instead of the MPEG-DASH stream fixes this. I'm trying to make a pull request as soon as possible.

**Comment by mediaminister** [comment]
I think I found the cause, there is a subtitle track defined in the MPEG-DASH manifest and InputStream Adaptive tries to play this subtitle track, but VRT streaming server returns HTTP Error 415 when requesting the subtitle stream segments.

**Comment by dagwieers** [comment]
Interesting. Maybe IA shouldn't insist on trying this 10 times on a HTTP 415 error either. But a problem with the stream nevertheless.

**Comment by dagwieers** [comment]
I am looking at the manifest and comparing the VRT Radio 1 and Radio 2 manifests I see these differences, which includes the subtitle track.
```patch
--- radio1.mpd	2020-05-01 01:07:55.000000000 +0200
+++ radio2.mpd	2020-05-01 01:11:18.000000000 +0200
@@ -6,7 +6,7 @@
   xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 http://standards.iso.org/ittf/PubliclyAvailableStandards/MPEG-DASH_schema_files/DASH-MPD.xsd"
   type="dynamic"
   availabilityStartTime="1970-01-01T00:00:00Z"
-  publishTime="2020-04-30T23:07:55.962091Z"
+  publishTime="2020-04-30T23:11:19.189273Z"
   minimumUpdatePeriod="PT2S"
   timeShiftBufferDepth="PT15M"
   maxSegmentDuration="PT2S"
@@ -37,9 +37,9 @@
         timescale="48000"
         initialization="live-$RepresentationID$.dash"
         media="live-$RepresentationID$-$Time$.dash">
-        <!-- 2020-04-30T22:52:40.656000Z / 1588287160 - 2020-04-30T23:07:39.216000Z -->
+        <!-- 2020-04-30T22:56:02.120000Z / 1588287362 - 2020-04-30T23:11:02.600000Z -->
         <SegmentTimeline>
-          <S t="76237783711488" d="92160" r="467" />
+          <S t="76237793381760" d="92160" r="468" />
         </SegmentTimeline>
       </SegmentTemplate>
       <Representation
@@ -53,29 +53,6 @@
     </AdaptationSet>
     <AdaptationSet
       id="2"
-      group="3"
-      contentType="text"
-      lang="nl"
-      mimeType="application/mp4"
-      codecs="stpp"
-      startWithSAP="1">
-      <Role schemeIdUri="urn:mpeg:dash:role:2011" value="caption" />
-      <SegmentTemplate
-        timescale="1000"
-        initialization="live-$RepresentationID$.dash"
-        media="live-$RepresentationID$-$Time$.dash">
-        <!-- 2020-04-30T22:52:40.640000Z / 1588287160 - 2020-04-30T23:07:41.120000Z -->
-        <SegmentTimeline>
-          <S t="1588287160640" d="1920" r="468" />
-        </SegmentTimeline>
-      </SegmentTemplate>
-      <Representation
-        id="text_track_0_dut=1000"
-        bandwidth="1000">
-      </Representation>
-    </AdaptationSet>
-    <AdaptationSet
-      id="3"
       group="2"
       contentType="video"
       par="16:9"
@@ -93,9 +70,9 @@
         timescale="600"
         initialization="live-$RepresentationID$.dash"
         media="live-$RepresentationID$-$Time$.dash">
-        <!-- 2020-04-30T22:52:40.640000Z / 1588287160 - 2020-04-30T23:07:41.120000Z -->
+        <!-- 2020-04-30T22:56:02.120000Z / 1588287362 - 2020-04-30T23:11:02.600000Z -->
         <SegmentTimeline>
-          <S t="952972296384" d="1152" r="468" />
+          <S t="952972417272" d="1152" r="468" />
         </SegmentTimeline>
       </SegmentTemplate>
       <Representation
@@ -143,85 +120,6 @@
         bandwidth="3000000"
         width="1280"
         height="720"
-        codecs="avc1.4D401F"
-        scanType="progressive">
-      </Representation>
-    </AdaptationSet>
-    <AdaptationSet
-      id="4"
-      group="2"
-      contentType="video"
-      par="16:9"
-      minBandwidth="145000"
-      maxBandwidth="3000000"
-      maxWidth="1280"
-      maxHeight="720"
-      segmentAlignment="true"
-      sar="1:1"
-      frameRate="5/16"
-      mimeType="video/mp4"
-      startWithSAP="1"
-      maxPlayoutRate="80"
-      codingDependency="false">
-      <EssentialProperty
-        schemeIdUri="http://dashif.org/guidelines/trickmode"
-        value="3">
-      </EssentialProperty>
-      <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main" />
-      <SegmentTemplate
-        timescale="600"
-        initialization="live-$RepresentationID$.dash"
-        media="live-$RepresentationID$-$Time$.dash">
-        <!-- 2020-04-30T22:52:40.640000Z / 1588287160 - 2020-04-30T23:07:41.120000Z -->
-        <SegmentTimeline>
-          <S t="952972296384" d="1152" r="468" />
-        </SegmentTimeline>
-      </SegmentTemplate>
-      <Representation
-        id="video_track=145000(mode=trik)"
-        bandwidth="145000"
-        width="416"
-        height="234"
-        codecs="avc1.4D401E"
-        scanType="progressive">
-      </Representation>
-      <Representation
-        id="video_track=365000(mode=trik)"
-        bandwidth="365000"
-        width="480"
-        height="270"
-        codecs="avc1.4D401E"
-        scanType="progressive">
-      </Representation>
-      <Representation
-        id="video_track=730000(mode=trik)"
-        bandwidth="730000"
-        width="640"
-        height="360"
-        codecs="avc1.4D401E"
-        scanType="progressive">
-      </Representation>
-      <Representation
-        id="video_track=1100000(mode=trik)"
-        bandwidth="1100000"
-        width="764"
-        height="432"
-        codecs="avc1.4D401F"
-        scanType="progressive">
-      </Representation>
-      <Representation
-        id="video_track=2000000(mode=trik)"
-        bandwidth="2000000"
-        width="960"
-        height="540"
-        codecs="avc1.4D401F"
-        scanType="progressive">
-      </Representation>
-      <Representation
-        id="video_track=3000000(mode=trik)"
-        bandwidth="3000000"
-        width="1280"
-        height="720"
         codecs="avc1.4D401F"
         scanType="progressive">
       </Representation>
```

**Comment by dagwieers** [comment]
If you look at the above manifest, you see that AdaptationSet 2 has the following incompatible information:
```json
      contentType="text"
      lang="nl"
      mimeType="application/mp4"
      codecs="stpp"
```

So it is a subtitle track with an "application/mp4" mimeType. Whereas video tracks typically use "video/mp4" for MPEG-4 video.

https://stackoverflow.com/questions/9929940/correct-mime-type-for-mp4

**Comment by mediaminister** [comment]
No, `mimeType="application/mp4"` is totally fine for subtitles: page 178 in https://dash-industry-forum.github.io/docs/DASH-IF-IOP-v4.3.pdf

It's just that the Radio 1 webcam doesn't have subtitles available, so the MPD manifest is wrongly configured.

**Comment by dagwieers** [comment]
@mediaminister But HTTP 415 would still be a strange error for something that is not available?

**Comment by mediaminister** [comment]
Indeed strange, but it's possible that the server returns a wrong http error code.

**Comment by nielslaukens** [comment]
Hi,

Let me start with a disclaimer: I work for the team responsible for the VRT streaming infrastructure, but I'm not involved in the deep technicalities.

The problem as I understand it is indeed what you already figured out above: Radio 1 (unlike our other radio channels) sometimes includes (closed, i.e. non-burned-in) subtitles. To support this case, the manifest includes a subtitle track. But theses subtitles are not used all the time: most of the time, there are no subtitles present. We can't leave these out of the manifest, because the player would need a full reload if we do need subtitles.

Since the subtitle track is not (always) present, the origin can't provide the requested data, and returns the 415 status that you observed. From the point of view of our origin partner, this is expected behaviour.

The player (TheoPlayer) we use on our official sites/apps ("touchpoints") handles this 415 as meaning "no subtitles present", and starts to play anyway. The players you tested (ffmpeg, jwplayer) behave differently. (Note that I'm trying to avoid pointing fingers at who's right or wrong)

We obviously do not support players other than our own, but we're looking in to always adding a (possibly empty) subtitle track. If this can be done easily, we'll do that. Otherwise, you'll have to fix this on the player-side.

**Comment by dagwieers** [comment]
@nielslaukens Great to get feedback like this and understand the technical background. This will definitely help to get other projects fix this corner case. We already established that (any) player should handle this more gracefully regardless of whether this was an oversight or not. Thanks!

**Comment by mediaminister** [comment]
> We can't leave these out of the manifest, because the player would need a full reload if we do need subtitles.

> From the point of view of our origin partner, this is expected behaviour.

I understand, I'll investigate how to handle this better on our side.

> but we're looking in to always adding a (possibly empty) subtitle track. If this can be done easily, we'll do that.

I'm glad there is a will to improve.

**Comment by dagwieers** [comment]
Just as a matter of due diligence I reported the playback issues with JW Player and FlowPlayer upstream as well:
- JW Player: https://github.com/jwplayer/jwplayer/issues/3680
- FlowPlayer: https://github.com/flowplayer/flowplayer/issues/1461

**Comment by dagwieers** [comment]
The VRT Radio 1 stream no longer seems to include the subtitle track. But at the same time a fix in inputstream.adaptive has been merged that would improve the undesired effect of missing subtitles on Kodi playback.

We may want to remove the workaround as soon as IA has a new release (v2.4.5 and v2.5.7). Keeping this open until that time.

**Comment by dagwieers** [comment]
Our fix is now released in Kodi v2.3.4, but in the meantime the subtitle track is no longer in the manifest. So I think we can undo our workaround, and then close this ticket.
